cosmwasm/contracts/ntt-global-accountant/src/contract.rs

```diff
@@ -437,59 +465,181 @@ fn handle_accountant_governance_vaa(
-    };
 
-    let key = transfer::Key::new(
-        body.emitter_chain.into(),
-        TokenAddress::new(body.emitter_address.0),
-        body.sequence,
-    );
+        // !IMPORTANT! the amounts are NOT normalized among different chains / tokens,
+        // i.e. the same token belonging to the same locking hub, can have message sourced from one chain that uses 4 decimals "normalized",
+        // and another that uses 8... this is maxed at 8, but should be actually normalized to 8 for accounting purposes.
+        let data = transfer::Data {
+            amount: Uint256::from(message.manager_payload.payload.amount.denormalize(8)),
+            token_address: hub.1,
+            token_chain: hub.0,
+            recipient_chain: message.manager_payload.payload.to_chain.id,
+        };
+
+        let key = transfer::Key::new(
+            source_chain,
+            TokenAddress::new(body.emitter_address.0),
+            body.sequence,
+        );
+
+        let tx = Transfer {
+            key: key.clone(),
+            data,
+        };
+        let evt = accountant::commit_transfer(deps.branch(), tx)
+            .with_context(|| format!("failed to commit transfer for key {key}"))?;
+
+        PENDING_TRANSFERS.remove(deps.storage, key);
+
+        Ok(evt)
+    } else if prefix == EndpointInit::PREFIX {
+        // only process init messages for locking hubs, setting their hub mapping to themselves
+        let message = EndpointInit::deserialize(&payload)?;
+        if message.manager_mode == (ManagerMode::LOCKING as u8) {
+            let chain = body.emitter_chain.into();
+            let hub_key = ENDPOINT_TO_HUB.key((chain, sender));
+
+            if hub_key
+                .may_load(deps.storage)
+                .context("failed to load hub")?
+                .is_some()
+            {
+                bail!("hub entry already exists")
+            }
+            hub_key
+                .save(deps.storage, &(chain, sender))
+                .context("failed to save hub")?;
+            Ok(Event::new("RegisterHub")
+                .add_attribute("chain", chain.to_string())
+                .add_attribute("emitter_address", hex::encode(sender)))
+        } else {
+            bail!("ignoring non-locking NTT initialization")
+        }
+    } else if prefix == EndpointRegister::PREFIX {
+        // for ease of code assurances, all endpoints should register their hub first, followed by other peers
+        // this code will only add peers for which it can assure their hubs match so one less key can be loaded on transfers
+        let message = EndpointRegister::deserialize(&payload)?;
+
+        let peer_hub_key =
+            ENDPOINT_TO_HUB.key((message.endpoint_chain_id, message.endpoint_address.into()));
+
+        let peer_hub = peer_hub_key
+            .may_load(deps.storage)
+            .context("failed to load peer hub")?
+            .ok_or(ContractError::MissingHubRegistration)?;
```

Comment:
So this will fail early for any peer registration that isn't the chainid of the locking chain where the hub has been registered?

Reply:
the first registration for a spoke (submitted to the accountant contract), has to be with its hub. this is so that the contract can enforce that all spokes share the same hub. they do not get peered together otherwise